File: packages/csharp/src/Compact.cs
using System.Text;
using System.Web;
using NSec.Cryptography;

namespace BlueScroll.HumanAttestation;

/// <summary>
/// HAP Compact Format utilities for space-efficient serialization.
///
/// Format: HAP{version}.{id}.{method}.{to_name}.{to_domain}.{at}.{exp}.{iss}.{signature}
///
/// Note: Effort dimensions (cost, time, physical, energy) are NOT included in compact format.
/// Compact is for QR codes - minimal representation. Full claims in JWS include all dimensions.
///
/// Example:
/// HAP1.hap_abc123xyz456.ba_priority_mail.Acme%20Corp.acme%2Ecom.1706169600.1769241600.ballista%2Ejobs.MEUCIQDx...
/// </summary>
public static class Compact
{
    private static readonly DateTimeOffset UnixEpoch = new DateTimeOffset(1970, 1, 1, 0, 0, 0, TimeSpan.Zero);

    /// <summary>
    /// Encodes a field for compact format (URL-encode + encode dots).
    /// </summary>
    private static string EncodeCompactField(string value)
    {
        return HttpUtility.UrlEncode(value ?? "").Replace(".", "%2E");
    }

    /// <summary>
    /// Decodes a compact format field.
    /// </summary>
    private static string DecodeCompactField(string value)
    {
        return HttpUtility.UrlDecode(value);
    }

    /// <summary>
    /// Convert ISO 8601 timestamp to Unix epoch seconds.
    /// </summary>
    private static long IsoToUnix(string iso)
    {
        var dt = DateTimeOffset.Parse(iso);
        return (long)(dt - UnixEpoch).TotalSeconds;
    }

    /// <summary>
    /// Convert Unix epoch seconds to ISO 8601 timestamp.
    /// </summary>
    private static string UnixToIso(long unix)
    {
        return UnixEpoch.AddSeconds(unix).ToString("O");
    }

    /// <summary>
    /// Encodes a HAP claim and signature into compact format (9 fields).
    /// </summary>
    /// <param name="claim">The claim to encode</param>
    /// <param name="signature">The Ed25519 signature bytes (64 bytes)</param>
    /// <returns>Compact format string</returns>
    public static string EncodeCompact(Claim claim, byte[] signature)
    {
        var name = claim.To?.Name ?? "";
        var domain = claim.To?.Domain ?? "";

        var atUnix = IsoToUnix(claim.At);
        var expUnix = string.IsNullOrEmpty(claim.Exp) ? 0 : IsoToUnix(claim.Exp);

        return string.Join(".",
            $"HAP{HumanAttestation.CompactVersion}",
            claim.Id,
            claim.Method ?? "",
            EncodeCompactField(name),
            EncodeCompactField(domain),
            atUnix.ToString(),
            expUnix.ToString(),
            EncodeCompactField(claim.Iss),
            Signer.Base64UrlEncode(signature)
        );
    }

    /// <summary>
    /// Decodes a compact format string into claim and signature.
    /// </summary>
    /// <param name="compact">The compact format string</param>
    /// <returns>DecodedCompact with claim and signature</returns>
    /// <exception cref="ArgumentException">If format is invalid</exception>
    public static DecodedCompact DecodeCompact(string compact)
    {
        if (!HumanAttestation.IsValidCompact(compact))
        {
            throw new ArgumentException("Invalid HAP Compact format");
        }

        var parts = compact.Split('.');
        if (parts.Length != 9)
        {
            throw new ArgumentException($"Invalid HAP Compact format: expected 9 fields, got {parts.Length}");
        }

        var version = parts[0];
        var hapId = parts[1];
        var method = parts[2];
        var encodedName = parts[3];
        var encodedDomain = parts[4];
        var atUnixStr = parts[5];
        var expUnixStr = parts[6];
        var encodedIss = parts[7];
        var sigB64 = parts[8];

        if (version != $"HAP{HumanAttestation.CompactVersion}")
        {
            throw new ArgumentException($"Unsupported compact version: {version}");
        }

        var name = DecodeCompactField(encodedName);
        var domain = DecodeCompactField(encodedDomain);
        var iss = DecodeCompactField(encodedIss);
        var atUnix = long.Parse(atUnixStr);
        var expUnix = long.Parse(expUnixStr);
        var signature = Signer.Base64UrlDecode(sigB64);

        var at = UnixToIso(atUnix);
        var exp = expUnix != 0 ? UnixToIso(expUnix) : null;

        var claim = new Claim
        {
            V = HumanAttestation.Version,
            Id = hapId,
            Method = method,
            Description = "", // Not included in compact format
            To = new ClaimTarget { Name = name },
            At = at,
            Iss = iss,
            Exp = exp
        };

        if (!string.IsNullOrEmpty(domain))
        {
            claim.To.Domain = domain;
        }

        return new DecodedCompact { Claim = claim, Signature = signature };
    }

    /// <summary>
    /// Builds the compact payload (everything before the signature).
    /// This is what gets signed.
    /// </summary>
    /// <param name="claim">The claim</param>
    /// <returns>Compact payload string (8 fields)</returns>
    public static string BuildCompactPayload(Claim claim)
    {
        var name = claim.To?.Name ?? "";
        var domain = claim.To?.Domain ?? "";

        var atUnix = IsoToUnix(claim.At);
        var expUnix = string.IsNullOrEmpty(claim.Exp) ? 0 : IsoToUnix(claim.Exp);

        return string.Join(".",
            $"HAP{HumanAttestation.CompactVersion}",
            claim.Id,
            claim.Method ?? "",
            EncodeCompactField(name),
            EncodeCompactField(domain),
            atUnix.ToString(),
            expUnix.ToString(),
            EncodeCompactField(claim.Iss)
        );
    }

    /// <summary>
    /// Signs a claim and returns it in compact format.
    /// </summary>
    /// <param name="claim">The claim to sign</param>
    /// <param name="privateKey">The Ed25519 private key</param>
    /// <returns>Signed compact format string</returns>
    public static string SignCompact(Claim claim, Key privateKey)
    {
        var payload = BuildCompactPayload(claim);
        var payloadBytes = Encoding.UTF8.GetBytes(payload);

        var algorithm = SignatureAlgorithm.Ed25519;
        var signature = algorithm.Sign(privateKey, payloadBytes);

        return $"{payload}.{Signer.Base64UrlEncode(signature)}";
    }

    /// <summary>
    /// Verifies a compact format string using provided public keys.
    /// </summary>
    /// <param name="compact">The compact format string</param>
    /// <param name="publicKeys">Array of JWK public keys to try</param>
    /// <returns>CompactVerificationResult</returns>
    public static CompactVerificationResult VerifyCompact(string compact, IEnumerable<Jwk> publicKeys)
    {
        if (!HumanAttestation.IsValidCompact(compact))
        {
            return new CompactVerificationResult { Valid = false, Error = "Invalid compact format" };
        }

        try
        {
            // Split to get payload and signature
            var lastDot = compact.LastIndexOf('.');
            var payload = compact[..lastDot];
            var sigB64 = compact[(lastDot + 1)..];
            var signature = Signer.Base64UrlDecode(sigB64);
            var payloadBytes = Encoding.UTF8.GetBytes(payload);

            var algorithm = SignatureAlgorithm.Ed25519;

            // Try each public key
            foreach (var jwk in publicKeys)
            {
                try
                {
                    var xBytes = Signer.Base64UrlDecode(jwk.X);
                    var publicKey = PublicKey.Import(algorithm, xBytes, KeyBlobFormat.RawPublicKey);

                    // Verify signature
                    if (algorithm.Verify(publicKey, payloadBytes, signature))
                    {
                        var decoded = DecodeCompact(compact);
                        return new CompactVerificationResult { Valid = true, Claim = decoded.Claim };
                    }
                }
                catch
                {
                    // Try next key
                    continue;
                }
            }

            return new CompactVerificationResult { Valid = false, Error = "Signature verification failed" };
        }
        catch (Exception e)
        {
            return new CompactVerificationResult { Valid = false, Error = e.Message };
        }
    }

    /// <summary>
    /// Generates a verification URL with embedded compact claim.
    /// </summary>
    /// <param name="baseUrl">Base verification URL (e.g., "https://ballista.jobs/v")</param>
    /// <param name="compact">The compact format string</param>
    /// <returns>URL with compact claim in query parameter</returns>
    public static string GenerateVerificationUrl(string baseUrl, string compact)
    {
        return $"{baseUrl}?c={HttpUtility.UrlEncode(compact)}";
    }

    /// <summary>
    /// Extracts compact claim from a verification URL.
    /// </summary>
    /// <param name="url">The verification URL</param>
    /// <returns>Compact string or null if not found</returns>
    public static string? ExtractCompactFromUrl(string url)
    {
        try
        {
            var uri = new Uri(url);
            var query = HttpUtility.ParseQueryString(uri.Query);
            var compact = query["c"];

            if (!string.IsNullOrEmpty(compact) && HumanAttestation.IsValidCompact(compact))
            {
                return compact;
            }

            return null;
        }
        catch
        {
            return null;
        }
    }
}

File: packages/go/compact.go
package humanattestation

import (
	"crypto/ed25519"
	"encoding/base64"
	"fmt"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// encodeCompactField encodes a field for compact format (URL-encode + encode dots)
func encodeCompactField(value string) string {
	encoded := url.QueryEscape(value)
	return strings.ReplaceAll(encoded, ".", "%2E")
}

// decodeCompactField decodes a compact format field
func decodeCompactField(value string) (string, error) {
	return url.QueryUnescape(value)
}

// base64urlEncode encodes bytes to base64url without padding
func base64urlEncode(data []byte) string {
	return base64.RawURLEncoding.EncodeToString(data)
}

// base64urlDecode decodes base64url string with padding restoration
func base64urlDecode(data string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(data)
}

// isoToUnix converts ISO 8601 timestamp to Unix epoch seconds
func isoToUnix(iso string) (int64, error) {
	t, err := time.Parse(time.RFC3339, iso)
	if err != nil {
		return 0, err
	}
	return t.Unix(), nil
}

// unixToISO converts Unix epoch seconds to ISO 8601 timestamp
func unixToISO(unix int64) string {
	t := time.Unix(unix, 0).UTC()
	return t.Format(time.RFC3339)
}

// EncodeCompact encodes a HAP claim and signature into compact format (9 fields)
func EncodeCompact(claim *Claim, signature []byte) (string, error) {
	atUnix, err := isoToUnix(claim.At)
	if err != nil {
		return "", fmt.Errorf("failed to parse 'at' timestamp: %w", err)
	}

	expUnix := int64(0)
	if claim.Exp != "" {
		expUnix, err = isoToUnix(claim.Exp)
		if err != nil {
			return "", fmt.Errorf("failed to parse 'exp' timestamp: %w", err)
		}
	}

	fields := []string{
		"HAP" + CompactVersion,
		claim.ID,
		claim.Method,
		encodeCompactField(claim.To.Name),
		encodeCompactField(claim.To.Domain),
		strconv.FormatInt(atUnix, 10),
		strconv.FormatInt(expUnix, 10),
		encodeCompactField(claim.Iss),
		base64urlEncode(signature),
	}

	return strings.Join(fields, "."), nil
}

// DecodeCompact decodes a compact format string into claim and signature
func DecodeCompact(compact string) (*DecodedCompact, error) {
	if !IsValidCompact(compact) {
		return nil, fmt.Errorf("invalid HAP Compact format")
	}

	parts := strings.Split(compact, ".")
	if len(parts) != 9 {
		return nil, fmt.Errorf("invalid HAP Compact format: expected 9 fields, got %d", len(parts))
	}

	version := parts[0]
	hapID := parts[1]
	method := parts[2]
	encodedName := parts[3]
	encodedDomain := parts[4]
	atUnixStr := parts[5]
	expUnixStr := parts[6]
	encodedIss := parts[7]
	sigB64 := parts[8]

	if version != "HAP"+CompactVersion {
		return nil, fmt.Errorf("unsupported compact version: %s", version)
	}

	name, err := decodeCompactField(encodedName)
	if err != nil {
		return nil, fmt.Errorf("failed to decode name: %w", err)
	}

	domain, err := decodeCompactField(encodedDomain)
	if err != nil {
		return nil, fmt.Errorf("failed to decode domain: %w", err)
	}

	iss, err := decodeCompactField(encodedIss)
	if err != nil {
		return nil, fmt.Errorf("failed to decode issuer: %w", err)
	}

	atUnix, err := strconv.ParseInt(atUnixStr, 10, 64)
	if err != nil {
		return nil, fmt.Errorf("failed to parse 'at' timestamp: %w", err)
	}

	expUnix, err := strconv.ParseInt(expUnixStr, 10, 64)
	if err != nil {
		return nil, fmt.Errorf("failed to parse 'exp' timestamp: %w", err)
	}

	signature, err := base64urlDecode(sigB64)
	if err != nil {
		return nil, fmt.Errorf("failed to decode signature: %w", err)
	}

	at := unixToISO(atUnix)
	var exp string
	if expUnix != 0 {
		exp = unixToISO(expUnix)
	}

	claim := &Claim{
		V:           Version,
		ID:          hapID,
		Method:      method,
		Description: "", // Not included in compact format
		To: ClaimTarget{
			Name:   name,
			Domain: domain,
		},
		At:  at,
		Iss: iss,
	}

	if exp != "" {
		claim.Exp = exp
	}

	return &DecodedCompact{
		Claim:     claim,
		Signature: signature,
	}, nil
}

// IsValidCompact validates if a string is a valid HAP Compact format
func IsValidCompact(compact string) bool {
	return CompactRegex.MatchString(compact)
}

// BuildCompactPayload builds the compact payload (everything before the signature)
// This is what gets signed.
func BuildCompactPayload(claim *Claim) (string, error) {
	atUnix, err := isoToUnix(claim.At)
	if err != nil {
		return "", fmt.Errorf("failed to parse 'at' timestamp: %w", err)
	}

	expUnix := int64(0)
	if claim.Exp != "" {
		expUnix, err = isoToUnix(claim.Exp)
		if err != nil {
			return "", fmt.Errorf("failed to parse 'exp' timestamp: %w", err)
		}
	}

	fields := []string{
		"HAP" + CompactVersion,
		claim.ID,
		claim.Method,
		encodeCompactField(claim.To.Name),
		encodeCompactField(claim.To.Domain),
		strconv.FormatInt(atUnix, 10),
		strconv.FormatInt(expUnix, 10),
		encodeCompactField(claim.Iss),
	}

	return strings.Join(fields, "."), nil
}

// SignCompact signs a claim and returns it in compact format
func SignCompact(claim *Claim, privateKey ed25519.PrivateKey) (string, error) {
	payload, err := BuildCompactPayload(claim)
	if err != nil {
		return "", err
	}

	signature := ed25519.Sign(privateKey, []byte(payload))
	return payload + "." + base64urlEncode(signature), nil
}

// VerifyCompact verifies a compact format string using provided public keys
func VerifyCompact(compact string, publicKeys []JWK) *CompactVerificationResult {
	if !IsValidCompact(compact) {
		return &CompactVerificationResult{Valid: false, Error: "Invalid compact format"}
	}

	// Split to get payload and signature
	lastDot := strings.LastIndex(compact, ".")
	payload := compact[:lastDot]
	sigB64 := compact[lastDot+1:]

	signature, err := base64urlDecode(sigB64)
	if err != nil {
		return &CompactVerificationResult{Valid: false, Error: fmt.Sprintf("failed to decode signature: %v", err)}
	}

	// Try each public key
	for _, jwk := range publicKeys {
		xBytes, err := base64.RawURLEncoding.DecodeString(jwk.X)
		if err != nil {
			continue
		}

		publicKey := ed25519.PublicKey(xBytes)

		// Verify signature
		if ed25519.Verify(publicKey, []byte(payload), signature) {
			// Signature is valid, decode the claim
			decoded, err := DecodeCompact(compact)
			if err != nil {
				return &CompactVerificationResult{Valid: false, Error: fmt.Sprintf("failed to decode claim: %v", err)}
			}
			return &CompactVerificationResult{Valid: true, Claim: decoded.Claim}
		}
	}

	return &CompactVerificationResult{Valid: false, Error: "Signature verification failed"}
}

// GenerateVerificationURL generates a verification URL with embedded compact claim
func GenerateVerificationURL(baseURL string, compact string) string {
	return baseURL + "?c=" + url.QueryEscape(compact)
}

// ExtractCompactFromURL extracts compact claim from a verification URL
func ExtractCompactFromURL(urlStr string) string {
	parsed, err := url.Parse(urlStr)
	if err != nil {
		return ""
	}

	compact := parsed.Query().Get("c")
	if compact != "" && IsValidCompact(compact) {
		return compact
	}

	return ""
}
